File: The_Algorithm_Design_Manual/4_Sorting_and_Searching/insertion_sort.c
/************************************************************/
/* File: insertion_sort.c                                   */
/* Insertion sort.                                          */
/************************************************************/

#include <stdio.h>

void insertionSort(int* arr, int size) {
    int i;
    for (i=2; i<size; i++) {
        int j=i;
        while (j>0 && arr[j] < arr[j-1]) {
            int tmp = arr[j];
            arr[j] = arr[j-1];
            arr[j-1] = tmp;
            j--;
        }
    }
}

int main(int argc, char* argv[]) {
    #define SIZE 7

    int arr[SIZE] = {2, 3, 1, 10, 8, 7, 9};
    insertionSort(arr, SIZE);

    int i;
    for (i=0; i<SIZE; i++) {
        printf("%d ", arr[i]);
    }
    printf("\n");

    return 0;
}

File: The_Algorithm_Design_Manual/4_Sorting_and_Searching/InsertionSort.java
public class InsertionSort {
    public static void sort(int[] arr) {
        for (int i=1; i<arr.length; i++) {
            int j = i;
            while (j>0 && arr[j] < arr[j-1]) {
                int tmp = arr[j-1];
                arr[j-1] = arr[j];
                arr[j] = tmp;
                j--;
            }
        }
    }

    public static void main(String args[]) {
        int[] arr = new int[] { 2,3,1,9,8,4,2 };
        InsertionSort.sort(arr);

        for (int i=0; i<arr.length; i++) {
            System.out.print(arr[i] + " ");
        }
        System.out.println();
    }
}
